File: app/src/main/java/com/drdisagree/colorblendr/common/Const.java
package com.drdisagree.colorblendr.common;

import com.drdisagree.colorblendr.BuildConfig;
import com.drdisagree.colorblendr.config.RPrefs;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Set;

public class Const {

    // Preferences file
    public static final String SharedPrefs = BuildConfig.APPLICATION_ID + "_preferences";

    // System packages
    public static final String FRAMEWORK_PACKAGE = "android";
    public static final String SYSTEMUI_PACKAGE = "com.android.systemui";
    public static final String SHELL_PACKAGE = "com.android.shell";

    // General preferences
    public static final String FIRST_RUN = "firstRun";
    public static final String THEMING_ENABLED = "themingEnabled";
    public static final String MONET_STYLE = "customMonetStyle";
    public static final String MONET_ACCENT_SATURATION = "monetAccentSaturationValue";
    public static final String MONET_BACKGROUND_SATURATION = "monetBackgroundSaturationValue";
    public static final String MONET_BACKGROUND_LIGHTNESS = "monetBackgroundLightnessValue";
    public static final String MONET_ACCURATE_SHADES = "monetAccurateShades";
    public static final String MONET_PITCH_BLACK_THEME = "monetPitchBlackTheme";
    public static final String MONET_SEED_COLOR = "monetSeedColor";
    public static final String MONET_SEED_COLOR_ENABLED = "monetSeedColorEnabled";
    public static final String MANUAL_OVERRIDE_COLORS = "manualOverrideColors";
    public static final String MONET_LAST_UPDATED = "monetLastUpdated";
    public static final String MONET_STYLE_ORIGINAL_NAME = "monetStyleOriginalName";
    public static final String FABRICATED_OVERLAY_SOURCE_PACKAGE = FRAMEWORK_PACKAGE;
    public static final String FABRICATED_OVERLAY_NAME_SYSTEM = BuildConfig.APPLICATION_ID + "_dynamic_theme";
    public static final String FABRICATED_OVERLAY_NAME_SYSTEMUI = BuildConfig.APPLICATION_ID + "_dynamic_theme_system";
    public static final String FABRICATED_OVERLAY_NAME_APPS = BuildConfig.APPLICATION_ID + ".%s_dynamic_theme";
    public static final String WALLPAPER_COLOR_LIST = "wallpaperColorList";
    public static final String FABRICATED_OVERLAY_FOR_APPS_STATE = "fabricatedOverlayForAppsState";
    public static final String SHOW_PER_APP_THEME_WARN = "showPerAppThemeWarn";
    public static final String TINT_TEXT_COLOR = "tintTextColor";
    public static final int SHIZUKU_PERMISSION_REQUEST_ID = 100;
    public static final String THEME_CUSTOMIZATION_OVERLAY_PACKAGES = "theme_customization_overlay_packages";
    public static final String SHIZUKU_THEMING_ENABLED = "shizukuThemingEnabled";

    // Service preferences
    public static final Gson GSON = new Gson();
    public static final String PREF_WORKING_METHOD = "workingMethod";

    public static Set<String> EXCLUDED_PREFS_FROM_BACKUP = new HashSet<>(
            Arrays.asList(
                    PREF_WORKING_METHOD,
                    MONET_LAST_UPDATED
            )
    );

    public static void saveSelectedFabricatedApps(HashMap<String, Boolean> selectedApps) {
        RPrefs.putString(FABRICATED_OVERLAY_FOR_APPS_STATE, Const.GSON.toJson(selectedApps));
    }

    public static HashMap<String, Boolean> getSelectedFabricatedApps() {
        String selectedApps = RPrefs.getString(FABRICATED_OVERLAY_FOR_APPS_STATE, null);
        if (selectedApps == null || selectedApps.isEmpty()) {
            return new HashMap<>();
        }

        return Const.GSON.fromJson(selectedApps, new TypeToken<HashMap<String, Boolean>>() {
        }.getType());
    }

    // Working method of app
    public enum WORK_METHOD {
        NULL,
        ROOT,
        SHIZUKU;

        public static WORK_METHOD fromString(String str) {
            try {
                return valueOf(str);
            } catch (Exception e) {
                return NULL;
            }
        }
    }

    public static WORK_METHOD WORKING_METHOD = WORK_METHOD.NULL;

    public static WORK_METHOD getWorkingMethod() {
        return WORK_METHOD.fromString(RPrefs.getString(PREF_WORKING_METHOD, WORK_METHOD.NULL.toString()));
    }

    public static void saveWorkingMethod(WORK_METHOD workMethod) {
        RPrefs.putString(PREF_WORKING_METHOD, workMethod.toString());
    }
}

File: app/src/main/java/com/drdisagree/colorblendr/common/Const.kt
package com.drdisagree.colorblendr.common

import com.drdisagree.colorblendr.BuildConfig
import com.drdisagree.colorblendr.config.RPrefs
import com.drdisagree.colorblendr.utils.SystemUtil
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

object Const {
    // Preferences file
    private const val OWN_PACKAGE_NAME = BuildConfig.APPLICATION_ID
    const val SHARED_PREFS = "${OWN_PACKAGE_NAME}_preferences"

    // Package names
    const val FRAMEWORK_PACKAGE = "android"
    const val SYSTEMUI_PACKAGE = "com.android.systemui"
    const val SHELL_PACKAGE = "com.android.shell"
    const val SYSTEMUI_CLOCKS = "com.android.systemui.clocks."
    const val GOOGLE_FEEDS = "com.google.android.googlequicksearchbox"
    const val GOOGLE_NEWS = "com.google.android.apps.magazines"
    const val PLAY_GAMES = "com.google.android.play.games"
    const val SETTINGS = "com.android.settings"
    const val SETTINGS_SEARCH = "com.google.android.settings.intelligence"
    const val PIXEL_LAUNCHER = "com.google.android.apps.nexuslauncher"

    // General preferences
    const val FIRST_RUN = "firstRun"
    const val THEMING_ENABLED = "themingEnabled"
    const val MONET_STYLE = "customMonetStyle"
    const val CUSTOM_MONET_STYLE = "userGeneratedMonetStyle"
    const val MODE_SPECIFIC_THEMES = "modeSpecificThemes"
    const val SCREEN_OFF_UPDATE_COLORS = "screenOffUpdateColors"
    const val DARKER_LAUNCHER_ICONS = "darkerLauncherIcons"
    const val SEMI_TRANSPARENT_LAUNCHER_ICONS = "semiTransparentLauncherIcons"
    const val FORCE_PITCH_BLACK_SETTINGS = "forcePitchBlackSettings"
    private val modeSpecificThemes: Boolean
        get() = RPrefs.getBoolean(MODE_SPECIFIC_THEMES, false)
    val MONET_ACCENT_SATURATION: String
        get() = if (!modeSpecificThemes) {
            "monetAccentSaturationValue"
        } else {
            if (SystemUtil.isDarkMode) "monetAccentSaturationValue" else "monetAccentSaturationValueLight"
        }
    val MONET_BACKGROUND_SATURATION: String
        get() = if (!modeSpecificThemes) {
            "monetBackgroundSaturationValue"
        } else {
            if (SystemUtil.isDarkMode) "monetBackgroundSaturationValue" else "monetBackgroundSaturationValueLight"
        }
    val MONET_BACKGROUND_LIGHTNESS: String
        get() = if (!modeSpecificThemes) {
            "monetBackgroundLightnessValue"
        } else {
            if (SystemUtil.isDarkMode) "monetBackgroundLightnessValue" else "monetBackgroundLightnessValueLight"
        }
    const val MONET_ACCURATE_SHADES = "monetAccurateShades"
    const val MONET_PITCH_BLACK_THEME = "monetPitchBlackTheme"
    const val MONET_SEED_COLOR_ENABLED = "monetSeedColorEnabled"
    const val MONET_SEED_COLOR = "monetSeedColor"
    const val MONET_SECONDARY_COLOR = "monetSecondaryColor"
    const val MONET_TERTIARY_COLOR = "monetTertiaryColor"
    const val MANUAL_OVERRIDE_COLORS = "manualOverrideColors"
    const val MONET_LAST_UPDATED = "monetLastUpdated"
    const val MONET_STYLE_ORIGINAL_NAME = "monetStyleOriginalName"
    const val FABRICATED_OVERLAY_SOURCE_PACKAGE = FRAMEWORK_PACKAGE
    const val FABRICATED_OVERLAY_NAME_SYSTEM = "${OWN_PACKAGE_NAME}_dynamic_theme"
    const val FABRICATED_OVERLAY_NAME_SYSTEMUI = "${OWN_PACKAGE_NAME}_dynamic_theme_system"
    const val FABRICATED_OVERLAY_NAME_APPS = "$OWN_PACKAGE_NAME.%s_dynamic_theme"
    const val WALLPAPER_COLOR_LIST = "wallpaperColorList"
    private const val FABRICATED_OVERLAY_FOR_APPS_STATE = "fabricatedOverlayForAppsState"
    const val SHOW_PER_APP_THEME_WARN = "showPerAppThemeWarn"
    const val TINT_TEXT_COLOR = "tintTextColor"
    const val SHIZUKU_PERMISSION_REQUEST_ID = 100
    const val THEME_CUSTOMIZATION_OVERLAY_PACKAGES = "theme_customization_overlay_packages"
    const val SHIZUKU_THEMING_ENABLED = "shizukuThemingEnabled"
    const val APP_LIST_FILTER_METHOD = "appListFilterMethod"
    const val SAVED_CUSTOM_MONET_STYLES = "savedCustomMonetStyles"

    // Service preferences
    val GSON: Gson = Gson()
    private const val PREF_WORKING_METHOD = "workingMethod"

    var EXCLUDED_PREFS_FROM_BACKUP: Set<String> = HashSet(
        listOf(
            FIRST_RUN,
            PREF_WORKING_METHOD,
            MONET_LAST_UPDATED,
            THEMING_ENABLED,
            SHIZUKU_THEMING_ENABLED,
            WALLPAPER_COLOR_LIST
        )
    )

    fun saveSelectedFabricatedApps(selectedApps: HashMap<String, Boolean>) {
        RPrefs.putString(FABRICATED_OVERLAY_FOR_APPS_STATE, GSON.toJson(selectedApps))
    }

    val selectedFabricatedApps: HashMap<String, Boolean>
        get() {
            val selectedApps = RPrefs.getString(
                FABRICATED_OVERLAY_FOR_APPS_STATE,
                null
            )
            if (selectedApps.isNullOrEmpty()) {
                return HashMap()
            }

            return GSON.fromJson(
                selectedApps,
                object :
                    TypeToken<HashMap<String?, Boolean?>?>() {
                }.type
            )
        }

    var WORKING_METHOD: WorkMethod = WorkMethod.NULL

    val workingMethod: WorkMethod
        get() = WorkMethod.fromString(
            RPrefs.getString(
                PREF_WORKING_METHOD,
                WorkMethod.NULL.toString()
            )
        )

    fun saveWorkingMethod(workMethod: WorkMethod) {
        RPrefs.putString(PREF_WORKING_METHOD, workMethod.toString())
    }

    // Working method of app
    enum class WorkMethod {
        NULL,
        ROOT,
        SHIZUKU;

        companion object {
            fun fromString(str: String?): WorkMethod {
                return try {
                    valueOf(str!!)
                } catch (e: Exception) {
                    NULL
                }
            }
        }
    }

    enum class AppType {
        SYSTEM,
        USER,
        LAUNCHABLE,
        ALL
    }
}
